File: calculator.py
import math


def add(x, y):
    return x + y

def subtract(x, y):
    return x - y

def multiply(x, y):
    return x * y

def divide(x, y):
    if y == 0:
        return "Cannot divide by zero!"
    return x / y


def sin(x):
    return math.sin(math.radians(x))

def cos(x):
    return math.cos(math.radians(x))

def tan(x):
    return math.tan(math.radians(x))


def log(x, base):
    return math.log(x, base)


def mean(data):
    return sum(data) / len(data)


def calculator():
    print("Welcome to the Advanced Calculator\n")
    print("Select operation:")
    print("1. Add")
    print("2. Subtract")
    print("3. Multiply")
    print("4. Divide")
    print("5. Sine")
    print("6. Cosine")
    print("7. Tangent")
    print("8. Logarithm")
    print("9. Mean")

    while True:
        choice = input("Enter choice (1-9): ")

        if choice in ['1', '2', '3', '4']:
            num1 = float(input("Enter first number: "))
            num2 = float(input("Enter second number: "))
            if choice == '1':
                print(f'The result is: {add(num1, num2)}')
            elif choice == '2':
                print(f'The result is: {subtract(num1, num2)}')
            elif choice == '3':
                print(f'The result is: {multiply(num1, num2)}')
            elif choice == '4':
                print(f'The result is: {divide(num1, num2)}')

        elif choice in ['5', '6', '7']:
            angle = float(input("Enter angle in degrees: "))
            if choice == '5':
                print(f'The sine of {angle} degrees is: {sin(angle)}')
            elif choice == '6':
                print(f'The cosine of {angle} degrees is: {cos(angle)}')
            elif choice == '7':
                print(f'The tangent of {angle} degrees is: {tan(angle)}')

        elif choice == '8':
            num = float(input("Enter number: "))
            base = float(input("Enter base: "))
            print(f'The logarithm of {num} base {base} is: {log(num, base)}')

        elif choice == '9':
            data = list(map(float, input("Enter numbers separated by space: ").split()))
            print(f'The mean is: {mean(data)}')
        else:
            print("Invalid input")

        next_calculation = input("Do you want to perform another calculation? (yes/no): ")
        if next_calculation.lower() != 'yes':
            break


if __name__ == '__main__':
    calculator()

File: calculator.js
class Calculator {
    // Basic Arithmetic Operations
    add(a, b) {
        return a + b;
    }

    subtract(a, b) {
        return a - b;
    }

    multiply(a, b) {
        return a * b;
    }

    divide(a, b) {
        if (b === 0) {
            throw new Error("Cannot divide by zero");
        }
        return a / b;
    }

    // Trigonometric Operations (in degrees)
    sin(angleDegrees) {
        return Math.sin(this.toRadians(angleDegrees));
    }

    cos(angleDegrees) {
        return Math.cos(this.toRadians(angleDegrees));
    }

    tan(angleDegrees) {
        return Math.tan(this.toRadians(angleDegrees));
    }

    arcsin(value) {
        if (value < -1 || value > 1) {
            throw new Error("arcsin input must be between -1 and 1");
        }
        return this.toDegrees(Math.asin(value));
    }

    arccos(value) {
        if (value < -1 || value > 1) {
            throw new Error("arccos input must be between -1 and 1");
        }
        return this.toDegrees(Math.acos(value));
    }

    arctan(value) {
        return this.toDegrees(Math.atan(value));
    }

    // Logarithmic Operations
    log(value, base = 10) {
        if (value <= 0) {
            throw new Error("Logarithm input must be positive");
        }
        if (base <= 0 || base === 1) {
            throw new Error("Logarithm base must be positive and not equal to 1");
        }
        return Math.log(value) / Math.log(base);
    }

    ln(value) {
        if (value <= 0) {
            throw new Error("Logarithm input must be positive");
        }
        return Math.log(value);
    }

    // Power and Root Operations
    power(base, exponent) {
        return Math.pow(base, exponent);
    }

    squareRoot(value) {
        if (value < 0) {
            throw new Error("Cannot take square root of negative number");
        }
        return Math.sqrt(value);
    }

    nthRoot(value, n) {
        if (n === 0) {
            throw new Error("Root degree cannot be zero");
        }
        if (value < 0 && n % 2 === 0) {
            throw new Error("Cannot take even root of negative number");
        }
        return Math.pow(value, 1 / n);
    }

    // Statistical Operations
    mean(numbers) {
        if (numbers.length === 0) {
            throw new Error("Cannot calculate mean of empty array");
        }
        return numbers.reduce((a, b) => a + b, 0) / numbers.length;
    }

    median(numbers) {
        if (numbers.length === 0) {
            throw new Error("Cannot calculate median of empty array");
        }
        const sorted = [...numbers].sort((a, b) => a - b);
        const mid = Math.floor(sorted.length / 2);
        return sorted.length % 2 !== 0 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
    }

    standardDeviation(numbers) {
        if (numbers.length < 2) {
            throw new Error("Need at least 2 numbers for standard deviation");
        }
        const avg = this.mean(numbers);
        const squareDiffs = numbers.map(x => Math.pow(x - avg, 2));
        const avgSquareDiff = squareDiffs.reduce((a, b) => a + b, 0) / (numbers.length - 1);
        return Math.sqrt(avgSquareDiff);
    }

    variance(numbers) {
        if (numbers.length < 2) {
            throw new Error("Need at least 2 numbers for variance");
        }
        const avg = this.mean(numbers);
        const squareDiffs = numbers.map(x => Math.pow(x - avg, 2));
        return squareDiffs.reduce((a, b) => a + b, 0) / (numbers.length - 1);
    }

    // Utility Operations
    factorial(n) {
        if (n < 0) {
            throw new Error("Factorial not defined for negative numbers");
        }
        if (n === 0 || n === 1) return 1;
        let result = 1;
        for (let i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }

    absoluteValue(value) {
        return Math.abs(value);
    }

    percentage(value, percent) {
        return (value * percent) / 100;
    }

    // Helper functions
    toRadians(degrees) {
        return degrees * (Math.PI / 180);
    }

    toDegrees(radians) {
        return radians * (180 / Math.PI);
    }
}

// Node.js export
if (typeof module !== "undefined" && module.exports) {
    module.exports = Calculator;
}